### PR TITLE
NP-793: Add new flag to Networks.operator.openshift.io for SDN live migration

### PR DESCRIPTION
**Pull request by pliurh**
Add flag 'mode' to 'spec.Migration' to indicate SDN migration mode.
The supported values are Live, Offline. If unset, the default mode is
Offline.

Implement https://github.com/openshift/enhancements/pull/1064

### CODE REVIEW

**Review by trozet — APPROVED** [review]
/approve

**Comment by trozet** [comment]
defer to @jcaamano for lgtm

File: openapi/openapi.json
Comment (jcaamano) [reply openapi/openapi.json:26999]: I think that might be for bool pointers specifically but not sure.

In any case, having something more explicit like `mode=live` might make more sense.

Comment (deads2k) [reply openapi/openapi.json:26999]: Correct, no booleans or pointers to booleans. `migrationMode` and `live` and whatever the alternative is, along with allowing "" and saying the platform chooses a default is a good way to go.

Comment (trozet) [reply openapi/openapi.json:26999]: migration from openshift-sdn -> ovnk

Comment (pliurh) [reply openapi/openapi.json:26999]: I changed the flag type. PTAL.

File: operator/v1/types_network.go
Comment (jcaamano) [inline operator/v1/types_network.go:154]: nits:

The possible values are usually quoted.

If you put it that way, why would one ever do an Offline migration? If there is no reason other than the fact that live is new an untested, maybe we could add a bit more detail in other differences about the two methods.

File: operator/v1/types_network.go
Comment (jcaamano) [inline operator/v1/types_network.go:126]: Why would you define a new type without exporting its possible values?
Would we have the alternative to inline the validation in the field instead of defining a new type?

File: operator/v1/types_network.go
Comment (pliurh) [reply operator/v1/types_network.go:154]: Yes, live is new and it will be only available for SD clusters. So we don't want to deprecate the offline one now. But in the long term, I think we will only support live migration.

File: operator/v1/types_network.go
Comment (pliurh) [reply operator/v1/types_network.go:126]: Oh, that's a mistake. I shall export the values. But defining a new type is a common practice in the API repo. e.g. https://github.com/openshift/api/blob/fd420c26854b53f62bd704b9cb8952a82c281403/machine/v1/types_controlplanemachineset.go#L87

And in the CRD file, we can see the validation is added as expected.

Comment (pliurh) [reply operator/v1/types_network.go]: ok, let's keep it simple. I changed the type to `string`.

File: operator/v1/types_network.go
Comment (JoelSpeed) [inline operator/v1/types_network.go:148]: Are there any interactions between this field and the other fields within this struct? Or in this type? Is this field compatible with all types of migrations currently allowed?

File: operator/v1/types_network.go
Comment (JoelSpeed) [reply operator/v1/types_network.go]: Preferred way is to do a defined type for enumerated values. Those which do not currently do this are not compliant with current API guidelines

**Comment by JoelSpeed** [comment]
@pliurh We are very close to the cut off for branching for 4.15. Do you have an implementation of this within the network operator that is ready to go and be merged once this API is ready? I think we should probably wait until after the 4.15 branch cut to merge this unless the implementation is ready to go. We do not want an API field present in 4.14 that has no implementation

**Comment by pliurh** [comment]
@JoelSpeed We will not have the implementation ready when 4.14 is GA. But we don't want to wait until 4.15 GA, we're trying to get it ready in 4.14.z. That is why we need this API change before the feature complete of 4.14.

File: operator/v1/types_network.go
Comment (pliurh) [reply operator/v1/types_network.go:140]: You're right. Users add this struct. CNO deploys the new CNI plugin, it will report the status in conditions of the network cluster operator CR. Users remove this struct once complete. CNO remove the old CNI plugin.

**Comment by JoelSpeed** [comment]
> @JoelSpeed We will not have the implementation ready when 4.14 is GA. But we don't want to wait until 4.15 GA, we're trying to get it ready in 4.14.z. That is why we need this API change before the feature complete of 4.14.

I've spoked to both @trozet and @zshi-redhat about this today, you need to get the feature complete in 4.15 and then you can backport the feature into a 4.14.z release. You should not merge an API if you think there's a chance it could ship without implementation.

File: operator/v1/types_network.go
Comment (pliurh) [reply operator/v1/types_network.go:148]: This new flag will be used together with the fields `NetworkType` and `Features`, but not `MTU`. Since MTU migration is a separate feature.

Comment (pliurh) [reply operator/v1/types_network.go:148]: If we drop `omitempty`, users have to configure this field whenever they configure the `spec.migration`. But the MTU migration doesn't need this field. Users only need to config `spec.migation.mtu` to trigger the MTU migration.

**Comment by pliurh** [comment]
@deads2k @JoelSpeed I added a tech preview feature gate for this feature. Do you think we can merge it before the 4.14 feature-complete?

File: operator/v1/types_network.go
Comment (JoelSpeed) [reply operator/v1/types_network.go]: Ok, so you have different kinds of migrations? What happens if a user were to set NetworkType and Features at the same time as MTU? Or MTU and Mode at the same time?

Comment (JoelSpeed) [reply operator/v1/types_network.go]: That's not true, if you drop `omitempty` but keep`""` in the enum, then this is a an optional field already, as denoted by `+optional`, so the field will be present, but will be `""`.

When was the MTU field added?

Comment (pliurh) [reply operator/v1/types_network.go]: I dropped `""`, and keep `omitempty` and `+optional`. Would it be ok?

The MTU field was added by https://github.com/openshift/api/pull/1078

Comment (JoelSpeed) [reply operator/v1/types_network.go]: Ok, it would make sense in that case to add a CEL validation that prevents the MTU and Mode fields being set at the same time, this would give more immediate feedback to the user

If you're putting these fields into tech preview, you'll want it like this, on the parent struct
```
// +openshift:validation:FeatureSetAwareXValidation:FeatureSet=CustomNoUpgrade;TechPreviewNoUpgrade,rule="!has(self.mtu) || !has(self.mode) || self.mode == ''",message="migration mode may not be configured at the same time as an mtu migration"
```

Comment (JoelSpeed) [reply operator/v1/types_network.go]: Can you raise a PR to add the `omitempty` and then test it with the CNO to see if the issue persists please

Comment (deads2k) [reply operator/v1/types_network.go]: > .spec.migration.mode: field not declared in schema

This suggests that the client didn't understand empty versus not set.  controller-runtime has a perma-bug doing this, but I'm not aware of the standard client having a problem.  Do you have a link to the client-call site?

**Pull request closed without merging**
